tests/unit/wordpress_mock.py

```diff
@@ -0,0 +1,526 @@
+        if not self.is_wordpress_installed(host, database):
+            raise KeyError(f"Wordpress isn't installed on (host={host}, database={database}).")
+        return self._databases[key]
+
+
+class MysqlConnectorMock:
+    """A mock for :py:mod:`mysql.connector`."""
+
+    Error = mysql.connector.Error
+
+    def __init__(self, wordpress_database_mock: WordpressDatabaseMock):
+        self._wordpress_database_mock = wordpress_database_mock
+
+    def connect(self, host: str, database: str, user: str, password: str, charset: str):
+        if not self._wordpress_database_mock.database_can_connect(host, database):
+            raise self.Error(
+                msg=f"Can't connect to MySQL server on '{host}:3306' (2003)",
+                errno=2003,
+            )
+        if not self._wordpress_database_mock.database_can_login(host, database, user, password):
+            raise self.Error(
+                msg=f"Access denied for user '{host}'@* (using password: *)",
+                errno=1045,
+            )
+
+        return unittest.mock.MagicMock()
+
+
+class HandlerRegistry:
+    """A utility class that can be used to collect pattern and handler pair using decorator syntax.
+
+    For example::
+        registry = HandlerRegistry()
+
+        @registry.register(match=lambda target: target.startswith("a"))
+        def handler_func(target):
+            print(target)
+
+        match, handler = registry.registered_handler[0]
+        match("abc") # => True
+        handler("abc") # => print("abc")
+    """
+
+    def __init__(self):
+        self.registered_handler = []
+
+    def register(
+        self, match: typing.Callable[[typing.Sequence[str]], bool]
+    ) -> typing.Callable[[typing.Callable], typing.Callable]:
+        """The decorator to collector the match pattern and handler, see class docstring for usage.
+
+        Args:
+            match: A match function takes input and output matching result as bool.
+        """
+
+        def decorator(func):
+            self.registered_handler.append((match, func))
+            return func
+
+        return decorator
+
+
+class ExecProcessMock:
+    """A mock for :class:`ops.pebble.ExecProcess`."""
+
+    def __init__(self, return_code: int, stdout: str, stderr: str):
+        self._return_code = return_code
+        self._stdout = stdout
+        self._stderr = stderr
+
+    def wait_output(self) -> typing.Tuple[str, str]:
+        if self._return_code != 0:
+            raise ops.pebble.ExecError(
+                [], exit_code=self._return_code, stdout=self._stdout, stderr=self._stderr
+            )
+        return self._stdout, self._stderr
+
+
+class WordpressContainerMock:
+    """A mock for :class:`ops.charm.model.Container`.
+
+    This will simulate file system and subprocess system inside the WordPress container.
+    """
+
+    _exec_handler = HandlerRegistry()
+
+    def __init__(
+        self,
+        wordpress_database_mock: WordpressDatabaseMock,
+    ):
```

Comment:
Should this have a return type and docstring?

Reply:
Fixed, thanks!